### DISCUSSION TITLE
What is the correct way to unmount / cleanup the components?

### DISCUSSION THREAD

**Discussion by KrishnaPG** [perspective-dev/perspective#1368] [Q&A] [answered]
In a single-page application (SPA) mounting and unmounting the Perspective on the fly requires cleaning up all the resources properly to avoid memory leaks and performance issues.

From the other issues and the documentation I was able to gather the below code sequence to be executed on unMount for a clean removal of the Perspective from the DOM without resource leak. Please confirm:

1. if any thing else need to be added for cleanup to ensure a clean unload
2. if the sequence order is correct

This assumes the perspective is loaded through CDN on the fly with script injection, and being unloaded:
````js
/*    
    worker = window.perspective.shared_worker();
    table =  worker.table(someSchema);
    viewerNode.load(table);
*/
cleanUp() {
   worker.terminate();              // stop the worker thread
   viewerNode.delete();           // delete for <perspective-viewer>
   table.delete();               // delete the data holder
   delete window.perspective; // remove the namespace

  //
  // now remove the custom components ['perspective-row', 'd3fc-canvas']
  //
  // however, custom components currently do not support unRegsiter / unDefine method
  // the best bet is to check if a custom component is already registered and avoid registering again.
  // Current no check is done and it results in errors such as shown in the below screenshot
}
````

![image](https://user-images.githubusercontent.com/4610831/108445654-c157be00-7282-11eb-8b60-6049a0916d63.png)

The goal is to be able to inject the `perspective` and unload it (mount, unMount cycles) **multiple times** on the fly without leaking resources or getting errors.

**Comment by texodus** [comment, marked as answer]
You have it mostly correct I think, but in the wrong order.
* `viewerNode.delete()` is the DOM element and must be deleted first, since it owns the `Table`.
*  `table.delete()` cleans up the memory associated with just this dataset and finally
* `worker.terminate()` kills the WebWorker itself, but you generally want to avoid doing this if you're going to start another worker later as they are expensive to start up and tear down, and effectively inert when unused.

Removing the namespace does nothing except make the library inaccessible - as you note there is not way AFAIK to de-register Custom Elements from the Browser once registered.  This is no more a leak than a symbol inaccessible allocated in a closure though (for example), as these data structures (the custom element class itself) are runtime static wrt memory usage.

**Reply by KrishnaPG** [reply]
Thank you @texodus
